jib-core/src/main/java/com/google/cloud/tools/jib/ncache/Cache.java

```diff
@@ -100,7 +106,26 @@ public CacheEntry write(Blob layerBlob, ImmutableList<LayerEntry> layerEntries,
       return Optional.empty();
     }
 
-    return cacheStorage.retrieve(optionalSelectedLayerDigest.get());
+    Optional<CacheEntry> optionalCacheEntry =
+        cacheStorage.retrieve(optionalSelectedLayerDigest.get());
+    if (!optionalCacheEntry.isPresent()) {
+      return Optional.empty();
+    }
+
+    CacheEntry cacheEntry = optionalCacheEntry.get();
+    if (!cacheEntry.getMetadataBlob().isPresent()) {
+      return Optional.empty();
+    }
+
+    Blob metadataBlob = cacheEntry.getMetadataBlob().get();
+    FileTime cacheEntryLastModifiedTime =
+        FileTime.from(Instant.parse(Blobs.writeToString(metadataBlob)));
```

Comment:
Hmm... so this assumes that this cache metadata `Blob` is an instance of `Instant`. I feel like `Blob` here is somewhat like the `Object` class in Java, obscuring anything inside it. Can't we have `CacheEntry.getLastSourceModifiedTime()`, much like we currently have `CacheEntry.getLayerSize()`, or have `getMetadata()` return a readable structure (i.e., return a Java class)?

Reply:
I intended `CacheEntry` to have metadata as a generic metadata whose contents would be determined by the user. `Cache` is basically the default API implementation for Jib builder to use, and defines the metadata as the last modified time of the layers (via its helper `LastModifiedMetadata`). Having `getMetadata` return `Blob` is intended to allow for different implementations of the metadata (like perhaps a fast file checksum or some other information).

Reply:
Then, is it possible to use generics somehow?

Reply:
Hmm, I think to have generics work, the cache retrieval (`DefaultCacheStorage`) would need to have knowledge of how to serialize/deserialize the metadata, which would either require the user providing a serialization/deserialization helper to `DefaultCacheStorage` or having `DefaultCacheStorage` use a specific serialization/deserialization method.

I think it might make sense to put this logic here in `LastModifiedMetadata`, like `LastModifiedMetadata#getLastModifiedTime(CacheEntry)`.

Reply:
Okay, I will leave it up to you, given that you have given enough thoughts on this.